tools/jxltran.cc

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+#include <string>
+#include <vector>
+
+#include "lib/include/jxl/decode.h"
+#include "lib/include/jxl/decode_cxx.h"
+#include "tools/cmdline.h"
+#include "tools/file_io.h"
+
+namespace jpegxl {
+namespace tools {
+namespace {
+
+struct Args {
+  void AddCommandLineOptions(CommandLineParser* cmdline) {
+    cmdline->AddPositionalOption("INPUT", /* required = */ true,
+                                 "The JPEG XL input file.", &file_in);
+
+    cmdline->AddPositionalOption("OUTPUT", /* required = */ true,
+                                 "The JPEG XL output file.", &file_out);
+
+    cmdline->AddHelpText("\nFile format options:", 0);
+
+    cmdline->AddOptionFlag('\0', "extract",
+                           "Extract the JPEG XL codestream"
+                           " from a file in the container file format.",
+                           &extract, &SetBooleanTrue);
+  }
+
+  const char* file_in = nullptr;
+  const char* file_out = nullptr;
+
+  bool extract = false;
+};
+
+JxlDecoderStatus extract_from_jxlc_box(
+    std::vector<uint8_t> const& input_bytes,
+    std::shared_ptr<std::vector<uint8_t>>& codestream_ptr) {
+
+  JxlDecoderPtr dec = JxlDecoderMake(nullptr);
+  if (!dec) {
+    fprintf(stderr, "JxlDecoderMake failed\n");
+    return JXL_DEC_ERROR ;
+  }
+
+  JxlDecoderStatus status = JxlDecoderSubscribeEvents(
+      dec.get(), JXL_DEC_BOX | JXL_DEC_BOX_COMPLETE);
+  if (status != JXL_DEC_SUCCESS) {
+    fprintf(stderr, "JxlDecoderSubscribeEvents failed\n");
+    return JXL_DEC_ERROR;
+  }
+
+  JxlDecoderSetInput(dec.get(), input_bytes.data(), input_bytes.size());
+  JxlDecoderCloseInput(dec.get());
+
+  static constexpr uint32_t kChunkSize = 65536;
+  codestream_ptr = std::make_shared<std::vector<uint8_t>>();
+  auto& codestream = *codestream_ptr;
+  uint32_t already_written_bytes = 0;
+
+  for (;;) {
+    status = JxlDecoderProcessInput(dec.get());
+    if (status == JXL_DEC_ERROR) {
+      fprintf(stderr, "Decoder error\n");
+      return JXL_DEC_ERROR;
+    } else if (status == JXL_DEC_NEED_MORE_INPUT) {
+      fprintf(stderr, "Error, already provided all input\n");
+      return JXL_DEC_ERROR;
+    } else if (status == JXL_DEC_BOX) {
+      JxlBoxType type;
+      status = JxlDecoderGetBoxType(dec.get(), type, /*decompressed=*/ false);
+      if (status != JXL_DEC_SUCCESS) {
+        fprintf(stderr, "Error, failed to get box type\n");
+        return JXL_DEC_ERROR;
+      }
+      if (!memcmp(type, "jxlc", 4)) {
```

Comment:
There is also the possibility of `jxlp` boxes, which contain part of the codestream and their contents has to be concatenated to get the full codestream. These boxes start with a uint32 index which has the msb set to 1 if it's the last `jxlp` box; the other 31 bits are a counter that is supposed to increment by one, in the same order as the order in which the boxes appear (you can just ignore it).

Reply:
Do you want me to do it right now? I'm usually working by doing the "smallest step possible", and support for `jxlp` boxes was planned as the very next PR.
And if your answer is yes, should I add a new commit or fixup the existing one?

Reply:
Can do it in a follow-up PR or in another commit in this one (they will be squashed when merging), both are fine.

Reply:
Will do in a follow-up then.